feat: Declarative TableMetadata Builder

c-thiel opened the pull request:
## Which issue does this PR close?

Currently we don't have a way to create `TableMetadata` in a declarative way other than deserializing it from JSON.
The `TableMetadataBuilder` mostly works imperatively with methods like `add_schema` and `add_partition_spec` that mutate the state incrementally. While incremental modification is what most users need, it is also helpful to offer a type safe way to create a new `TableMetadata` object given all of its attributes, other than deserialization from JSON.

My concrete use-case is: Lakekeeper stores TableMetadata in Postgres. We extract data typesafe, so we end up with various objects such as a list of schemas, partition_specs, the last_updated_ms and all other fields required to build a new `TableMetadata`. As we already have rust-types, we don't want to serialize them to JSON first. We also can't use the `TableMetadata` builder, as it's incremental nature would result in significant overhead.

## Desin considerations
Instead of using the builder approach, we could also just add another method `try_from_parts` or similar to `impl TableMetadata`. We have this alternative approach implemented here: https://github.com/lakekeeper/iceberg-rust/blob/a8b6509775b139c92772a999b0cbca637e274a03/crates/iceberg/src/spec/table_metadata.rs#L676-L740

I feel that a builder is less intrusive though. I don't like adding the `UnnormalizedTableMetadata`, but didn't find a way to skip this type without writing significantly more code that needs maintenance. Because we have the derived builder here now, this approach is almost maintenance free.

## What changes are included in this PR?

* Introduce a new derived `TableMetadataDeclarativeBuilder`
* Introduce `UnnormalizedTableMetadata` as an intermediate type that is `.try_normalize()` to `TableMetadata`

## Are these changes tested?

yes

c-thiel commented:
@liurenjie1024 @Xuanwo  Looking forward to your thoughts!

findepi commented on crates/iceberg/src/spec/table_metadata.rs:
```diff
@@ -172,10 +180,18 @@ pub struct TableMetadata {
+    }))]
     pub(crate) statistics: HashMap<i64, StatisticsFile>,
     /// Mapping of snapshot ids to partition statistics files.
+    #[builder(default)]
+    #[builder(setter(transform = |stats: Vec<PartitionStatisticsFile>| {
```
why not `#[builder(default, setter...`? (provided `default` stays, see above)

findepi commented on crates/iceberg/src/spec/table_metadata.rs:
```diff
@@ -172,10 +180,18 @@ pub struct TableMetadata {
     /// even if the refs map is null.
     pub(crate) refs: HashMap<String, SnapshotReference>,
     /// Mapping of snapshot ids to statistics files.
+    #[builder(default, setter(transform = |stats: Vec<StatisticsFile>| {
+        stats.into_iter().map(|s| (s.snapshot_id, s)).collect()
```
Why also default? How is this different from `snapshots`?
Accepting `Vec<StatisticsFile>` should be enough.

c-thiel replied:
Its not - I missed the default for `snapshots` and `current_snapshot_id` and a few others.
My basic reasoning is that all fields that must be set in order to produce valid metadata should not have a default, even if the Rust type would allow it. An example for a non-default field would be `schemas`, as we require at least one entry in the HashMap to be able to produce valid Metadata (`current_schema_id` must be set and be present).
Valid `TableMetadata` must not contain `snapshots` though, and `current_snapshot_id` can be None, so those should have defaults.

findepi commented on crates/iceberg/src/spec/table_metadata.rs:
```diff
@@ -3018,4 +3062,62 @@ mod tests {
+            .metadata_log(vec![])
+            .refs(HashMap::new())
+            .build()
+            .try_normalize()
```
In the test for the declarative builder would be better to avoid unnecessary calls. Do we need `metadata.try_normalize`? (the test passes without `metadata.try_normalize` call)

OTOH it's somewhat weird that `try_normalize` is the only API to finish building. 
Could `build()` do the `try_normalize` implicitly, so that user doesn't have to?

c-thiel replied:
I would prefer this design as well, however I can't find an easy way to implement a custom build method without implementing it on the very complex generic type that is derived by the macro (see end of this message) and is hard to keep in sync.

`typed_builder`s only usable customization of the build method is using `Into` to convert the type. There is no native way to inject `try_normalize` into the build method. Especially, there is no way to make the generated build method return a `Result`.

I see the following options:

1. Just build `TableMetadata` without normalization - this would allow to create a potentially invalid TableMetadata, as all other constructors (existing `TableMetadataBuilder` & `Deserialize`) use normalization. Hence, I discarded it.
2. Implement a fallible `try_build` (with normalize) method and make the derived `build` method private. This would require to implement the `try_build` on the above type, which isn't very obvious, and would need to be kept in sync with fields in `TableMetadata` and even worse, with which have a default builder implementation. Due to this complexity, and a high probability of state drift, I discarded this as well.
3. Don't use typed builder - which requires us to write significantly more code, but it would be a viable option.
4. Extend typed builder to support `TryInto` in addition to `Into` in the build method.
5. Use a very small custom type that wraps the notion of unvalidated metadata and offers an easy way to validate it - which is what I did.

If there are any other ideas here or if I missed something, let me know!

Here is the type: Note that default and non-default fields are treated differently:
```rust
#[allow(dead_code, non_camel_case_types, missing_docs)]
#[automatically_derived]
impl<
    __properties: ::typed_builder::Optional<HashMap<String, String>>,
    __current_snapshot_id: ::typed_builder::Optional<Option<i64>>,
    __snapshots: ::typed_builder::Optional<HashMap<i64, SnapshotRef>>,
    __snapshot_log: ::typed_builder::Optional<Vec<SnapshotLog>>,
    __metadata_log: ::typed_builder::Optional<Vec<MetadataLog>>,
    __statistics: ::typed_builder::Optional<HashMap<i64, StatisticsFile>>,
    __partition_statistics: ::typed_builder::Optional<HashMap<i64, PartitionStatisticsFile>>,
    __encryption_keys: ::typed_builder::Optional<HashMap<String, String>>,
>
    TableMetadataDeclarativeBuilder<(
        (FormatVersion,),
        (Uuid,),
        (String,),
        (i64,),
        (i64,),
        (i32,),
        (HashMap<i32, SchemaRef>,),
        (i32,),
        (HashMap<i32, PartitionSpecRef>,),
        (PartitionSpecRef,),
        (StructType,),
        (i32,),
        __properties,
        __current_snapshot_id,
        __snapshots,
        __snapshot_log,
        __metadata_log,
        (HashMap<i64, SortOrderRef>,),
        (i64,),
        (HashMap<String, SnapshotReference>,),
        __statistics,
        __partition_statistics,
        __encryption_keys,
    )>
```

findepi commented on crates/iceberg/src/spec/table_metadata.rs:
```diff
@@ -101,7 +101,10 @@ pub const RESERVED_PROPERTIES: [&str; 9] = [
-#[derive(Debug, PartialEq, Deserialize, Eq, Clone)]
+#[derive(Debug, PartialEq, Deserialize, Eq, Clone, typed_builder::TypedBuilder)]
+#[builder(builder_method(name=declarative_builder))]
+#[builder(builder_type(name=TableMetadataDeclarativeBuilder, doc="Build a new [`TableMetadata`] in a declarative way. For imperative operations (e.g. `add_snapshot`) and creating new TableMetadata, use [`TableMetadataBuilder`] instead."))]
```
I am not convinced "declarative" vs "imperative" is the defining distinction.
The `TableMetadataBuilder` is not "a builder pattern for TableMetadata". It's a facade to perform changes on the TableMetadata to be invoked by higher layer. For that reason, it includes _implicit changes_ that are supposed to be carried on when a metadata change happens, such as updating last_updated_ms when a new snapshot is added
https://github.com/apache/iceberg-rust/blob/3ea6f4791a2a4e15c4466cff7ccc478c08e2d3e3/crates/iceberg/src/spec/table_metadata_builder.rs#L394

The use-case described in the PR description is "the" standard builder pattern. To me, this is the defining difference. 

I suggest changing the doc string here and align the type names.

c-thiel replied:
I absolutely agree with this from a rust perspective. From an iceberg perspective other languages use the world `Builder` for this functionality unfortunately. I am uncertain if renaming the `TableMetadataBuilder` to a clearer name is worth the breaking change today.

findepi commented on crates/iceberg/src/spec/table_metadata.rs:
```diff
@@ -101,7 +101,10 @@ pub const RESERVED_PROPERTIES: [&str; 9] = [
 pub type TableMetadataRef = Arc<TableMetadata>;
 
-#[derive(Debug, PartialEq, Deserialize, Eq, Clone)]
+#[derive(Debug, PartialEq, Deserialize, Eq, Clone, typed_builder::TypedBuilder)]
```
nit: import typed_builder::TypedBuilder

findepi commented on crates/iceberg/src/spec/table_metadata.rs:
```diff
@@ -654,6 +670,33 @@ impl TableMetadata {
+
+impl UnnormalizedTableMetadata {
+    /// Build a new [`UnnormalizedTableMetadata`] using the [`TableMetadataDeclarativeBuilder`].
+    pub fn builder() -> TableMetadataDeclarativeBuilder {
```
unused? remove?

c-thiel replied:
This is a library and the constructor can be used outside of the crate. I like to attach some constructor to a type if available.
As it's an auxiliary type, I am OK with removing it as well.

ZENOTME commented on crates/iceberg/src/spec/table_metadata.rs:
```diff
@@ -121,20 +124,23 @@ pub struct TableMetadata {
     ///A string to string map of table properties. This is used to control settings that
     /// affect reading and writing and is not intended to be used for arbitrary metadata.
-    /// For example, commit.retry.num-retries is used to control the number of commit retries.
+    /// For example, commit.retry.num-retries is used to control the number of commit retries
```
nit: unnecessary change?

c-thiel commented:
@findepi @ZENOTME  thanks for the Reviews!
I addressed most of your comments in my latest commit.

The most unclear point is probably the return type of the Build method.
Getting more Feedback on my https://github.com/apache/iceberg-rust/pull/1362#discussion_r2140970629 would be great!

c-thiel commented:
@liurenjie1024 thanks for your review and sorry for letting this drag so long.
I switched now to a much simpler approach that just uses a `build_unchecked` method to construct the `TableMetadata`.
I believe this is by far the most maintainable, as it is just derived. I am ok with a breaking change in this builder if the `TableMetadata` object changes, in fact, I would expect it. We do have a stable builder already that works imperative. This builder is just here to construct this specific struct from its parts, so I think its ok that the builder changes if the struct changes.
